SEQNG-52: Implementation of engine startup options

jluhrs opened the pull request:
Read engine startup settings from configuration file app.cfg. The settings are then applied to the Seqexec engine. The implementation of the settings themselves is a bit dirty (uses mutability), but I think it is OK, has they are not supposed to change in run-time.
I also moved the code that configures the Seqexec engine inside ```SeqexecEngine```

cquiroz commented:
Could we make them immutable? In practice it is mostly converting the `object` to `class` and pass the settings as constructor parameters. You can see that on the configuration of the web app classes

jdnavarro commented:
In general, what sort of code should go into the engine module vs the server module (not the web server module)? I still can't see a clearcut criterion.

What you mean by immutable? Is it doing unsafe IO?

jdnavarro commented on modules/edu.gemini.seqexec.web/edu.gemini.seqexec.web.server/src/main/scala/edu/gemini/seqexec/web/server/http4s/WebServerLauncher.scala:
```diff
@@ -120,8 +102,8 @@ object WebServerLauncher extends ServerApp with LogInitialization {
-      sc <- executorConf
-      _  <- seqexecExecutor.run(sc)
+      c <- config
+      _  <- SeqexecEngine.seqexecConfigure.run(c)
```
Why not follow the existing pattern of configuration? Previously there was not need to do in-place mutations and wrap them into a `Task`. When running this you are just unwrapping the `Task` you just introduced in `seqexecConfigure`, but under the hood you already did an unsafe mutation outside of a `Task`.

I'm not saying this is outright wrong because maybe there is no way to around it, but I'm trying to understand what was the problem with the previous configuration pattern.

cquiroz commented on modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/SeqexecEngine.scala:
```diff
@@ -116,4 +118,22 @@ object SeqexecEngine {
+      val instSim = cfg.require[Boolean]("seqexec-engine.instSim")
+      val gcalSim = cfg.require[Boolean]("seqexec-engine.gcalSim")
+
+      Task.delay(ODBProxy.host(new Peer(odbHost, 8443, null))) *>
```
I guess this could be something like new `SeqTranslate(setting, dhsClient, odbProxy)` and return it on the `Kleisli`

jluhrs commented:
About the Execution vs Server separation, I have being trying to keep the logic about running a sequence in Engine, while keeping the interaction with external systems (like the ODB) in Server

jluhrs commented:
I think I addressed all the concerns with the last update. Instead of using objects with vars for singletons that needed initialization parameters, I use classes and pass an instance around.

cquiroz commented on modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/SeqTranslate.scala:
```diff
@@ -8,47 +10,56 @@ import edu.gemini.spModel.seqcomp.SeqConfigNames._
+                    flamingos2: Flamingos2Controller
+                    )
+
+  private val dhsSimulator = DhsClientSim(LocalDate.now)
```
@jdnavarro I guess this also falls into a side effect?

jdnavarro replied:
Yes, but it might not be worth introducing a `Task` with this PR. With a `TODO` comment to not forget about it in the future would be good enough.

cquiroz commented on modules/edu.gemini.seqexec.web/edu.gemini.seqexec.web.server/src/main/scala/edu/gemini/seqexec/web/server/http4s/WebServerLauncher.scala:
```diff
@@ -120,11 +102,11 @@ object WebServerLauncher extends ServerApp with LogInitialization {
       wc <- serverConf
-      sc <- executorConf
-      _  <- seqexecExecutor.run(sc)
+      c <- config
```
Just a nitpick, could we vertically align the for element?

jdnavarro commented on modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/SeqexecEngine.scala:
```diff
@@ -10,13 +10,23 @@ import scalaz._
   * Created by jluhrs on 10/7/16.
   */
-object SeqexecEngine {
+class SeqexecEngine(settings: SeqexecEngine.Settings) {
```
👍
